apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,231 @@
+  init: function ms_init() {
+    EVENTS.forEach((event) => {
+      mozMobileMessage.addEventListener(
+        event,
+        this[Utils.camelCase(`on-${event}`)]
+      );
+    });
+
+    METHODS.forEach((shimMethod) => {
+      service.method(shimMethod, this[shimMethod]);
+    });
+
+    STREAMS.forEach((shimStream) => {
+      service.stream(shimStream, this[shimStream]);
+    });
+  },
+
+  /* Events */
+
+  onSending: function ms_onSending(e) {
+    service.broadcast('message-sending', { message: e.message });
+  },
+
+  onFailed: function ms_onFailed(e) {
+    service.broadcast('message-failed-to-send', { message: e.message });
+  },
+
+  onDeliverysuccess: function ms_onDeliverysuccess(e) {
+    service.broadcast('message-delivered', { message: e.message });
+  },
+
+  onReadsuccess: function ms_onReadsuccess(e) {
+    service.broadcast('message-read', { message: e.message });
+  },
+
+  onSent: function ms_onSent(e) {
+    service.broadcast('message-sent', { message: e.message });
+  },
+
+  onReceived: function ms_onReceived(e) {
+    var message = e.message;
+
+    if (message.messageClass && message.messageClass === 'class-0') {
+      return;
+    }
+
+    // Here we can only have one sender, so deliveryInfo[0].deliveryStatus =>
+    // message status from sender. Ignore 'pending' messages that are received
+    // this means we are in automatic download mode
+    if (message.delivery === 'not-downloaded' &&
+        message.deliveryInfo[0].deliveryStatus === 'pending') {
+      return;
+    }
+
+    service.broadcast('message-received', { message: message });
+  },
+
+  onDeleted: function ms_onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Method */  
+
+  getMessage: function ms_getMessag(id) {
+    return mozMobileMessage.getMessage(id);
+  },
+
+  retrieveMMS: function ms_retrieveMMS(id) {
+    return mozMobileMessage.retrieveMMS(id);
+  },
+
+  send: function ms_send(recipient, content, sendOpts) {
+    return mozMobileMessage.send(recipient, content, sendOpts);
+  },
+
+  sendMMS: function ms_sendMMS(dataOpts, sendOpts) {
+    return mozMobileMessage.sendMMS(dataOpts, sendOpts);
+  },
+
+  delete: function ms_delete(id) {
+    return mozMobileMessage.delete(id);
+  },
+
+  markMessageRead: function ms_markMessageRead(id, isRead, returnReq) {
+    return mozMobileMessage.markMessageRead(id, isRead, returnReq);
+  },
+
+  getSegmentInfoForText: function ms_getSegmentInfoForText(text) {
+    return mozMobileMessage.getSegmentInfoForText(text);
+  },
+
+  /* Stream */
+
+  /**
+   * Stream wrapper for getThreads API for returning threads in chunck.
+   * @param {*} stream channel for returning thread and flag.
+   */
+
+  getThreads: function ms_getThreads(stream) {
+    var cursor = null;
+    var options = {
+      data: null,
+      end: false,
+      done: false
+    }
```

Comment:
yeah, I agree we should call close/abort for complete/error case here and the view(client) should handle in stream.closed promise